core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -637,9 +637,11 @@ private[spark] class BlockManager(
   def reregister(): Unit = {
     // TODO: We might need to rate limit re-registering.
     logInfo(s"BlockManager $blockManagerId re-registering with master")
-    master.registerBlockManager(blockManagerId, diskBlockManager.localDirsString, maxOnHeapMemory,
-      maxOffHeapMemory, storageEndpoint)
-    reportAllBlocks()
+    val id = master.registerBlockManager(blockManagerId, diskBlockManager.localDirsString,
+      maxOnHeapMemory, maxOffHeapMemory, storageEndpoint, isReRegister = true)
+    if (id.executorId != BlockManagerId.INVALID_EXECUTOR_ID) {
+      reportAllBlocks()
+    }
```

Comment:
Do we want to terminate in case of `INVALID_EXECUTOR_ID` ?

Reply:
Good question. So ideally, a lost executor should be terminated in the end anyways (whether killed by the driver or exit itself proactively)...if the lost executor fails to terminate, there must be something wrong with it. So I think terminate here won't make any difference to the result.

Reply:
If it is failing to terminate, every heartbeat will end up returning `BlockManagerId.INVALID_EXECUTOR_ID` from driver right ?
Wondering if there is a reason to keep it around.

Reply:
Note - if we do change this - it is introduction of a new code path.
So I dont want to block the PR on this discussion - but would be good to understand this case better :-)

Reply:
The case we met is that the executor failed to be killed by both `StopExecutor` and `ExecutorRunner.killProcess`. After second thinking, maybe killing (via `System.exit`) from inside would give a more chance to save this case?

Reply:
Added the termination logic here. cc @mridulm